2.0/m_geoipban.cpp

```diff
@@ -0,0 +1,109 @@
+/*
+ * InspIRCd -- Internet Relay Chat Daemon
+ *
+ *   Copyright (C) 2008 Robin Burchell <robin+git@viroteck.net>
+ *
+ * This file is part of InspIRCd.  InspIRCd is free software: you can
+ * redistribute it and/or modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation, version 2.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+#include "inspircd.h"
+#include <GeoIP.h>
+
+#ifdef _WIN32
+# pragma comment(lib, "GeoIP.lib")
+#endif
+
+/* $ModDesc: Implements extban +b G: - GeoIP Country Code bans */
+/* $LinkerFlags: -lGeoIP */
+
+class ModuleGeoIPBan : public Module
+{
+
+	LocalStringExt ext;
+	GeoIP* gi;
+
+	std::string* SetExt(User* user)
+	{
+		const char* c = GeoIP_country_code_by_addr(gi, user->GetIPString());
+		if (!c)
+			c = "UNK";
+
+		std::string* cc = new std::string(c);
+		ext.set(user, cc);
+		return cc;
+	}
+
+ public:
+	ModuleGeoIPBan() : ext("geoip_cc", this), gi(NULL)
+	{
+	}
+		
+ public:
+	void init()
+	{
+		gi = GeoIP_new(GEOIP_STANDARD);
+		if (gi == NULL)
+				throw ModuleException("Unable to initialize geoip, are you missing GeoIP.dat?");
+		ServerInstance->Modules->AddService(ext);		
+		Implementation eventlist[] = { I_OnCheckBan, I_On005Numeric, I_OnWhois};
+		ServerInstance->Modules->Attach(eventlist, this, sizeof(eventlist)/sizeof(Implementation));
+	}
+
+	~ModuleGeoIPBan()
+	{
+	}
+
+	Version GetVersion()
+	{
+		return Version("Extban 'G' - GeoIP Country Code ban", VF_OPTCOMMON|VF_VENDOR);
+	}
+
+	ModResult OnCheckBan(User *user, Channel *c, const std::string& mask)
+	{
+
+		std::string* cc = ext.get(user);
+		if (!cc)
+			cc = SetExt(user);
+
+		if ((mask.length() > 2) && (mask[0] == 'G') && (mask[1] == ':'))
+		{				
+			if (InspIRCd::Match(*cc, mask.substr(2)))
+				return MOD_RES_DENY;
+		}
+		return MOD_RES_PASSTHRU;
+	}
+
+	void On005Numeric(std::string &output)
+	{
+		ServerInstance->AddExtBanChar('G');
+	}
+
+	virtual void OnWhois(User* src, User* dst)
```

Comment:
yeah, it adds their country to their whois? which make normal user can see anyone country and that not good 
most of user don't want any find that ,, it's their rights " privacy " matter
but it will be good if it's work only for ircops